### ISSUE TITLE
combine gmpy logic in one place

### ISSUE THREAD

**Issue by haru-44** [sympy/sympy#25067]
see #25059

We would like to put together a declaration of the gmpy function in `sympy.external.gmpy`.
Since gmpy is optional, we need to provide an alternative function if gmpy is not used. The challenge is that similar functions are already used in several places. We would like to explore their use and move to `external` if possible.

(Work in progress)

- [x] gcd
  * functions similar to `gmpy2.gcd`
    + `math.gcd`
    + `sympy.core.numbers.igcd` : has a cache mechanism
    + `mpmath.libmp.gcd`
  * `math.gcd` may be used.
  * PR
    + #25077
    + #25080
- [x] lcm
  * functions similar to `gmpy2.lcm`
    + `math.lcm`
    + `sympy.core.numbers.ilcm`
  * PR
    + #25085
    + #25096
- [x] invert
  * functions similar to `gmpy2.invert`
    + `sympy.core.numbers.mod_inverse`
    + `sympy.core.numbers.igcdex` : often substituted for
    + `pow`
  * `pow` may be used. However, the difference is that `pow` raises a `ValueError` when the inverse element does not exist, while gmpy raises a `ZeroDivisionError`.
  * PR
    + #25083
- [x] iroot
  * functions similar to `gmpy2.iroot`
    + `sympy.core.power.integer_nthroot`
  * PR
    + #25504
- [x] isqrt
  * functions similar to `gmpy2.isqrt`
    + `sympy.core.power.isqrt`
    + `math.isqrt`
    + `mpmath.libmp.isqrt`
  * We currently use `mpmath`, but I think we could use `math`.
  * Fixed mpmath due to speed problems.
  * PR
    + #25111
    + mpmath/mpmath#695
- [x] is_square
  * functions similar to `gmpy2.is_square`
    + `sympy.ntheory.primetest.is_square`
  * PR
    + #25095
    + #25306
- [x] legendre, jacobi
  * functions similar to `gmpy2.legendre`
    + `sympy.ntheory.residue_ntheory.legendre_symbol`
  * functions similar to `gmpy2.jacobi`
    + `sympy.ntheory.residue_ntheory.jacobi_symbol`
  * PR
    + #25140

**Comment by Rayuga35** [comment]
I wanted to contribute to Algebra, but I didn't understand what is the issue here and what is this all about? If anybody can help me with this, I would appreciate it

**Comment by haru-44** [comment]
The `gcd` changes have been merged, but there was a problem.

`igcd` takes integers as arguments. The "integer" here could be `sympy.core.numbers.Integer` as well as the python built-in `int`. If you give `numbers.Integer` to `math.gcd` or `numbers.igcd`, it returns the answer as expected, but if you give `numbers.Integer` to `gmpy2.gcd`, you raise a `TypeError`.

To solve this problem, I am considering the following segregation.
* `sympy.external.gmpy.gcd` takes `int` or `mpz` as input.
* `numbers.igcd` takes `numbers.Integer` as input.

**Comment by oscarbenjamin** [comment]
> To solve this problem, I am considering the following segregation.
> 
> * `sympy.external.gmpy.gcd` takes `int` or `mpz` as input.
> * `numbers.igcd` takes `numbers.Integer` as input.

I think that it is better not to pass `Integer` to these functions so we should use `sympy.external.gmpy.gcd` everywhere internally but also not use `Integer` internally most of the time. An `Integer` can be converted to an `int` using its `.p` attribute. Any calculation that is not symbolic should either not accept `Integer` or should immediately convert `Integer` to `int` (or MPZ).

**Comment by smichr** [comment]
I left a comment on #25085

**Comment by oscarbenjamin** [comment]
Thanks @haru-44 

Perhaps a followup to this would be to add the same pieces for flint after gh-25474. Longer term I would like flint to be the default choice for ground types and there are many more pieces in Flint itself that are not yet exposed by python-flint but that SymPy would want to make use of.

An important advantage in having all this logic in one place is that it makes it easier to switch between multiple different implementations in different cases.

**Comment by asmeurer** [comment]
By the way, since we are working on this, do you any of you have any thoughts on how hard it would be to make the grounds types swappable at runtime? 

Right now they have to be set before SymPy is imported and cannot be changed. It would be make testing against different ground types easier if they didn't have to be tested in different processes, and if flint doesn't support everything it will be a necessity to be able to fallback to gmpy and I suspect this would make that easier. The downside of this is that you'd need to make sure that objects created with one ground type don't break down if they are combined with objects created with a different one.

**Comment by oscarbenjamin** [comment]
> do you any of you have any thoughts on how hard it would be to make the grounds types swappable at runtime?

It would be very fiddly I think. There are things like `ZZ_python` and `ZZ_gmpy` that you can set interchangeably in a situation where you set the domain explicitly. Very little code sets the domain explicitly though and if it does then it uses e.g. `ZZ`. Changing `construct_domain` to return a different version of `ZZ` is one thing but having the `ZZ` object be a single object that dynamically chooses whether it is supposed to be `ZZ_python` or `ZZ_gmpy` would be tricky. The `ZZ` object is stored in lots of other objects and its `__eq__` would have to change behaviour depending on the dynamic ground types which would invalidate all hash tables etc.

Basically it could be done but it would be difficult and would be flakey. Probably if the capability was there then we would not want to recommend that anyone uses it.

> if flint doesn't support everything it will be a necessity to be able to fallback to gmpy

Flint has vastly more than the libraries that gmpy2 wraps so long term I doubt that there will be any situation where we would use Flint as the ground types but still want to use any gmpy2 functions. Note that Flint actually includes gmpy2's dependencies GMP and MPFR as its own dependencies so any feature of GMP can be trivially a feature of Flint. SymPy does not use gmpy2's MPFR/MPC features anywhere and neither does mpmath as far as I know.
